### PR TITLE
[pytato, compile] Transform DAG before making it a placeholder

### PR DESCRIPTION
**Pull request by kaushikcfd**
Transforming is generally helpful as it may reveal new metadata information that wasn't available earlier.

### CODE REVIEW

File: arraycontext/impl/pytato/compile.py [hunk 1]
```diff
@@ -154,7 +154,34 @@ def id_collector(keys, ary):
...
+                                                              TaggableCLArray)
+    if isinstance(ary, pt.Array):
+        dag = pt.make_dict_of_named_arrays({"_actx_out": ary})
+        return actx.transform_dag(dag)["_actx_out"].expr
+    elif isinstance(ary, TaggableCLArray):
+        return ary
+    elif isinstance(ary, cla.Array):
+        from warnings import warn
+        warn("Passing pyopencl.array.Array to the compiled callable"
+             " will be deprecated in 2023."
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:173]: "is deprecated and will stop working"?

File: arraycontext/impl/pytato/compile.py [hunk 2]
```diff
@@ -154,7 +154,34 @@ def id_collector(keys, ary):
...
+    from arraycontext.impl.pyopencl.taggable_cl_array import (to_tagged_cl_array,
+                                                              TaggableCLArray)
+    if isinstance(ary, pt.Array):
+        dag = pt.make_dict_of_named_arrays({"_actx_out": ary})
+        return actx.transform_dag(dag)["_actx_out"].expr
+    elif isinstance(ary, TaggableCLArray):
+        return ary
+    elif isinstance(ary, cla.Array):
+        from warnings import warn
+        warn("Passing pyopencl.array.Array to the compiled callable"
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:172]:
```suggestion
        warn("Passing pyopencl.array.Array to a compiled callable"
```

File: arraycontext/impl/pytato/compile.py [hunk 3]
```diff
@@ -154,7 +154,34 @@ def id_collector(keys, ary):
     return pmap(arg_id_to_arg), pmap(arg_id_to_descr)
 
 
-def _get_f_placeholder_args(arg, kw, arg_id_to_name):
+def _transform_array(ary, actx):
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:157]: Not loving the name. `_to_input_for_compiled`?

File: arraycontext/impl/pytato/compile.py [hunk 4]
```diff
@@ -154,7 +154,34 @@ def id_collector(keys, ary):
...
+    """
+    Returns the transformed version of *ary* after calling
+    :meth:`ArrayContext.transform_dag` on it.
+    """
+    import pyopencl.array as cla
+    from arraycontext.impl.pyopencl.taggable_cl_array import (to_tagged_cl_array,
+                                                              TaggableCLArray)
+    if isinstance(ary, pt.Array):
+        dag = pt.make_dict_of_named_arrays({"_actx_out": ary})
+        return actx.transform_dag(dag)["_actx_out"].expr
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:167]: Add a comment that this is about metadata propagation.

File: arraycontext/impl/pytato/compile.py [hunk 5]
```diff
@@ -154,7 +154,34 @@ def id_collector(keys, ary):
     return pmap(arg_id_to_arg), pmap(arg_id_to_descr)
 
 
-def _get_f_placeholder_args(arg, kw, arg_id_to_name):
+def _transform_array(ary, actx):
+    """
+    Returns the transformed version of *ary* after calling
+    :meth:`ArrayContext.transform_dag` on it.
+    """
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:161]: The docstring doesn't totally reflect all the types that this function accepts, or why it's useful.

File: arraycontext/impl/pytato/compile.py [hunk 6]
```diff
@@ -165,10 +192,16 @@ def _get_f_placeholder_args(arg, kw, arg_id_to_name):
         return pt.make_placeholder(name, (), np.dtype(type(arg)))
     elif isinstance(arg, pt.Array):
         name = arg_id_to_name[(kw,)]
-        return pt.make_placeholder(name, arg.shape, arg.dtype)
+        # Transform the DAG to learn about any new metadata
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:195]:
```suggestion
        # Transform the DAG to give metadata inference a chance to do its job
```

File: arraycontext/impl/pytato/compile.py [hunk 7]
```diff
@@ -165,10 +192,16 @@ def _get_f_placeholder_args(arg, kw, arg_id_to_name):
...
+        arg = _transform_array(arg, actx)
+        return pt.make_placeholder(name, arg.shape, arg.dtype,
+                                   axes=arg.axes,
+                                   tags=arg.tags)
     elif is_array_container_type(arg.__class__):
         def _rec_to_placeholder(keys, ary):
             name = arg_id_to_name[(kw,) + keys]
+            # Transform the DAG to learn about any new metadata
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:203]:
```suggestion
            # Transform the DAG to give metadata inference a chance to do its job
```

**Review by inducer — COMMENTED** [review]
Thanks! A few concerns regarding code readability, otherwise good to go.

File: arraycontext/impl/pytato/compile.py (hunk 1 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/compile.py]: Done, thanks!

File: arraycontext/impl/pytato/compile.py (hunk 2 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/compile.py]: Done, thanks!

File: arraycontext/impl/pytato/compile.py (hunk 3 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/compile.py]: Yep, that's better. Fixed!

File: arraycontext/impl/pytato/compile.py (hunk 4 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/compile.py:174]: Done!

File: arraycontext/impl/pytato/compile.py (hunk 5 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/compile.py]: Added more details to the docs. Thanks!

File: arraycontext/impl/pytato/compile.py (hunk 6 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/compile.py]: Thanks!

File: arraycontext/impl/pytato/compile.py (hunk 7 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/compile.py]: Thanks!

**Comment by inducer** [comment]
Thanks!

**Pull request merged**